Mathlib/Algebra/Lie/CartanExists.lean

```diff
@@ -0,0 +1,380 @@
+import Mathlib.Algebra.Lie.CartanSubalgebra
+import Mathlib.Algebra.Lie.Rank
+import Mathlib.LinearAlgebra.Eigenspace.Zero
+
+/-!
+# Existence of Cartan subalgebras
+
+In this file we prove existence of Cartan subalgebras in finite-dimensional Lie algebras,
+following [barnes1967].
+
+## Main results
+
+* `exists_isCartanSubalgebra_of_finrank_le_card`:
+  A Lie algebra `L` over a field `K` has a Cartan subalgebra,
+  provided that the dimension of `L` over `K` is less than or equal to the cardinality of `K`.
+* `exists_isCartanSubalgebra`:
+  A finite-dimensional Lie algebra `L` over an infinite field `K` has a Cartan subalgebra.
+
+## References
+
+* [barnes1967]: "On Cartan subalgebras of Lie algebras" by D.W. Barnes.
+
+-/
+
+namespace LieAlgebra
+
+section CommRing
+
+variable {K R L M : Type*}
+variable [Field K] [CommRing R] [Nontrivial R]
+variable [LieRing L] [LieAlgebra K L] [LieAlgebra R L]
+variable [AddCommGroup M] [Module R M] [LieRingModule L M] [LieModule R L M]
+variable [Module.Finite K L]
+variable [Module.Finite R L] [Module.Free R L]
+variable [Module.Finite R M] [Module.Free R M]
+
+open FiniteDimensional LieSubalgebra Module.Free Polynomial
+
+variable (K)
+
+lemma finrank_engel (x : L) :
+    finrank K (engel K x) = (ad K L x).charpoly.natTrailingDegree :=
+  (ad K L x).finrank_maximalGeneralizedEigenspace
+
+lemma rank_le_finrank_engel (x : L) :
+    rank K L ≤ finrank K (engel K x) :=
+  (rank_le_natTrailingDegree_charpoly_ad K x).trans
+    (finrank_engel K x).ge
+
+lemma isRegular_iff_finrank_engel_eq_rank (x : L) :
+    IsRegular K x ↔ finrank K (engel K x) = rank K L := by
+  rw [isRegular_iff_natTrailingDegree_charpoly_eq_rank, finrank_engel]
+
+namespace engel_le_engel
+
+/-!
+## Implementation details for the proof of `LieAlgebra.engel_le_engel`
+
+In this section we provide some auxiliary definitions and lemmas
+that are used in the proof of `LieAlgebra.engel_le_engel`,
+which is the following statement:
+
+Let `L` be a Lie algebra of dimension `n` over a field `K` with at least `n` elements.
+Given a Lie subalgebra `U` of `L`, and an element `x ∈ U` such that `U ≤ engel K x`.
+Suppose that `engel K x` is minimal amongst the Engel subalgebras `engel K y` for `y ∈ U`.
+Then `engel K x ≤ engel K y` for all `y ∈ U`.
+
+We follow the proof strategy of Lemma 2 in [barnes1967].
+-/
+
+variable (R M)
+
+variable (x y : L)
+
+open LieModule LinearMap
+
+local notation "φ" => LieModule.toEndomorphism R L M
+
+/-- Let `x` and `y` be elements of a Lie `R`-algebra `L`, and `M` a Lie module over `M`.
+Then the characteristic polynomials of the family of endomorphisms `⁅r • y + x, _⁆` of `M`
+have coefficients that are polynomial in `r : R`.
+In other words, we obtain a polynomial over `R[X]`
+that specializes to the characteristic polynomial of `⁅r • y + x, _⁆` under the map `X ↦ r`.
+This polynomial is captured in `lieCharpoly R M x y`. -/
+private noncomputable
+def lieCharpoly : Polynomial R[X] :=
```

Comment:
```suggestion
def lieCharpoly : R[X][X] :=
```
Or maybe even better if we used the [PolynomialPolynomial scope](https://github.com/leanprover-community/mathlib4/blob/8d0c6e6f821392660468fdce7c5463740d7988be/Mathlib/AlgebraicGeometry/EllipticCurve/Affine.lean#L84) then this could be `R[X][Y]`. (This would require moving that scope to a more basic file.)